Experiment: add smoke tests to P2 frontend editor using Playwright.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR is an **experiment** to add smoke test coverage to the frontend P2 editor, as requested in #56664.

Key changes:
- set up a P2 test blog, visible to external users (e2eflowtesting users are technically treated as external users).
- implement a TOTP client to permit 2FA (required for any user accessing P2s).
- new class `IsolatedBlockEditorComponent` to interact with non-iframed Gutenberg editor on P2s.
- new spec under `test/e2e/specs/specs-p2` to log in, write a post and ensure post is visible.

#### Testing instructions

No TeamCity build configuration is set to run this spec at this moment.

To test:

1. pull the branch locally.
2. navigate to `test/e2e`.
3. run `yarn jest --verbose specs/specs-p2/p2-post.ts
`yarn jest --verbose specs/specs-p2/p2-post.ts`

The output should be something like:

```
 PASS  specs/specs-p2/p2-post.ts (20.509 s)
  Post
    ✓ Log In (6211 ms)
    ✓ View P2 (3403 ms)
    ✓ Add a Paragraph block (874 ms)
    ✓ Enter text (214 ms)
    ✓ Submit post (165 ms)
    ✓ Validate post submission was successful (5925 ms)

Test Suites: 1 passed, 1 total
Tests:       6 passed, 6 total
Snapshots:   0 total
Time:        20.558 s, estimated 50 s
Ran all test suites matching /specs\/specs-p2\/p2-post.ts/i.
```

Remaining todo items:
- [ ] Set up TeamCity build configuration for P2 tests.
- [ ] File issue for refactoring LoginPage.login and LoginPage.enter2FA.
- [ ] File issue for refactoring GutenbergEditorPage and IsolatedBlockEditorComponent.

Related to #56664.

worldomonation committed:
Add p2User to config file.

worldomonation committed:
Implement TOTP client for `@automattic/calypso-e2e`.

worldomonation committed:
Implement smoke test for P2.

worldomonation commented:
@dpasque Pinging you for an initial look. What do you think about how everything is laid out?

worldomonation commented on packages/calypso-e2e/src/lib/blocks/paragraph-block.ts:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Represents the Audio block.
+ */
+export class ParagraphBlock {
```
This block isn't strictly necessary for a smoke test of the frontend editor, but at the same time I figured might as well implement one.

My opinion was that it didn't need to be perfect, just do what it needs to do (which is to enter text into the block).

worldomonation commented on packages/calypso-e2e/src/lib/components/isolated-block-editor-component.ts:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Represents an instance of the WPCOM's Gutenberg editor page.
+ */
+export class IsolatedBlockEditorComponent {
```
I took this name from https://github.com/Automattic/isolated-block-editor/ which appears to be the name for anytime Gutenberg editor is used but not within the context of a WordPress.com site. So in this case, a P2 frontend editor matches this description.

worldomonation commented on packages/calypso-e2e/src/lib/components/isolated-block-editor-component.ts:
```diff
@@ -0,0 +1,56 @@
+	 * 	Typically built into the definition file for the block being inserted.
+	 * @returns {ElementHandle} Reference to the inserted block in the editor.
+	 */
+	async addBlock( blockName: string, blockEditorSelector: string ): Promise< ElementHandle > {
+		// Click on the editor title. This has the effect of dismissing the block inserter
+		// if open, and restores focus back to the editor root container, allowing insertion
+		// of blocks.
+		await this.page.click( selectors.blockInserterToggle );
+		await this.page.fill( selectors.blockInserterSearch, blockName );
+		await this.page.click( `${ selectors.blockInserterResultItem } span:text("${ blockName }")` );
+		// Confirm the block has been added to the editor body.
+		return await this.page.waitForSelector( `${ blockEditorSelector }.is-selected` );
```
I have identified the need to merge this with GutenbergEditorComponent, or at least create a base/superset of the two similar editor classes. 

The `addBlock` methods in both `GutenbergEditorPage` and `IsolatedBlockEditorComponent` can be essentially identical, minus the use of `Page` instead of `Frame` in one case versus the other. Perhaps this is an ideal location to re-introduce (in limited capacity) the notion of inheritance?

As this is an experimental PR meant to demonstrate that we can indeed run e2e tests on P2s, I've chosen to essentially copy-paste the `addBlock` method, but once this is merged I think its' worth exploring how we can combine the two.

dpasque replied:
`Perhaps this is an ideal location to re-introduce (in limited capacity) the notion of inheritance?`

My personal preference would be strongly to stay away from inheritance if possible! We've avoided it so far, and I just always find it makes things messier... But that's just my 0.02!

Perhaps instead we could make a `Flow` class to wrap this function, and that class takes in its constructor either a `page` or `frame`, and then the function just uses whatever was passed in in that constructor?

We could still then expose these methods on the POM classes, they would just call the shared function in the Flow class!

worldomonation replied:
> Perhaps instead we could make a Flow class to wrap this function, and that class takes in its constructor either a page or frame

That's a good idea! It would allow unifying of the API without having to alter the existing files. We'd have to update the test specs to use the new flow, but that isn't too big of a deal.

One other avenue that occurred to me is to maintain the separation. The reasoning behind this is as follows:
- IsolatedBlockEditor is maintained in a separate repository, so the POM for this should also live as a separate item.
- GutenbergEditorPage is concerned with the full-page editing experience (iframed or not), so it's a separate entity from the front-end, (slightly) simplified editing experience present on P2s.

In any case, perhaps this exploration should be done alongside adding Atomic support to GutenbergEditorPage.

dpasque replied:
>One other avenue that occurred to me is to maintain the separation.

I think that's totally valid too!

>We'd have to update the test specs to use the new flow

Actually, you wouldn't necessarily have to! Although it's slightly inverted than usual, you could actually have the existing methods in the existing POM classes call the Flow class, and then keep the existing scripts referencing the POM classes!

worldomonation commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -265,4 +269,30 @@ export class LoginPage {
+		await this.page.fill( selectors.mfaInput, code );
+
+		const [ response ] = await Promise.all( [
+			this.page.waitForResponse( '**/wp-login.php?action=two-step-authentication-endpoint' ),
```
This endpoint is slightly different from the one used in `LoginPage.login`, but the overall outline of the method is the same.

In a subsequent PR (not this one due to scope creep) I think it's worth looking at ways to extract the commonalities into a separate method.

worldomonation commented on packages/calypso-e2e/src/lib/pages/p2-page.ts:
```diff
@@ -0,0 +1,28 @@
+	 *
+	 */
+	async focusInlineEditor(): Promise< void > {
+		await this.page.click( selectors.inputPrompt );
```
I've treated this method as being separate from the `IsolatedBlockEditorComponent` due to the primary highlight being on the P2 page itself when first landing on a P2 site.

dpasque commented on packages/calypso-e2e/src/lib/blocks/paragraph-block.ts:
```diff
@@ -0,0 +1,49 @@
+};
+
+/**
+ * Represents the Audio block.
```
Small nit on comment here! It's a Paragraph block 😄

dpasque commented on packages/calypso-e2e/src/totp-client.ts:
```diff
@@ -0,0 +1,24 @@
+/**
+ *
+ */
+export class TOTPClient {
+	private secret: string;
+
+	/**
+	 *
+	 * @param secret
+	 */
+	constructor( secret: string ) {
+		this.secret = secret;
+	}
+
+	/**
+	 *
+	 * @returns {string} TOTP code valid for the epoch for which it was generated.
+	 */
+	getToken(): string {
+		return totp( this.secret ).toString();
+	}
+}
```
Totally optional, but do you see this functionality growing significantly?

If this class just will be used to generate a token, what do you think about making just a globally exported library function that takes the secret at the first parameter? Again, a totally optional style thing, just wanted to throw it out there!

worldomonation replied:
Functionally I do not see this class growing at all.

With that said, I think it might be worth keeping it as a class to draw parallel to the `EmailClient` which is in its own class. I realize the scope of its actions is vastly different, but conceptually it makes more sense (to me) that TOTP is implemented as a separate client that needs to be instantiated.

dpasque commented on test/e2e/specs/specs-p2/p2-post.ts:
```diff
@@ -0,0 +1,68 @@
+	} );
+
+	it( 'Validate post submission was successful', async function () {
+		await page.waitForSelector( `.entry-content:has-text("${ postContent }")` );
```
Small nit for when we actually do this, but I think we could move this to a function somewhere on the P2 or isolated editor POM!

worldomonation replied:
I will move this into the P2Page.

dpasque reviewed (commented):
I think this looks fantastic, thanks for spinning this up!! 🎉   Got nothing but some small nits, but I know this is just a draft so feel free to just ignore those if not helpful :)

worldomonation committed:
Review comments:
- move validation of post to P2Page.
- add JSDoc to classes.
- fix block name in JSDoc.

worldomonation committed:
Rename suite name

dpasque reviewed (approved):
Looks great, I'm loving this smoke test! 🎉

worldomonation committed:
update yarn.lock

worldomonation committed:
Fix rebase errors in config file and test spec.
